walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  /**
+   * Finds available time ranges for all requested attendees to attend the
+   * supplied MeetingRequest, given known events
+   * 
+   * @param events  All known events that will be considered
+   * @param request The meeting request whose attendees must all be free
+   * @return A collection of time ranges where all proposed attendees in @request
+   *         are not attending any known events in @events
+   */
+  public Collection<TimeRange> query(final Collection<Event> events, final MeetingRequest request) {
+    Collection<TimeRange> queryRanges = new ArrayList<>();
+    queryRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    Map<Event, Boolean> cache = new HashMap<>();
+    for (Event event : events) {
+      Collection<TimeRange> nextRanges = new ArrayList<>();
+      for (TimeRange range : queryRanges) {
+        if (hasOverlappingAttendees(event, request, cache)) {
+          nextRanges.addAll(splitToAvailableTimeRanges(range, event.getWhen()));
+        } else {
+          nextRanges.add(range);
+        }
+      }
+      queryRanges = nextRanges;
+    }
+    List<TimeRange> resultRanges = new ArrayList<>();
+    for (TimeRange range : queryRanges) {
+      if (range.duration() >= request.getDuration()) {
+        resultRanges.add(range);
+      }
+    }
+    Collections.sort(resultRanges, TimeRange.ORDER_BY_END);
+    return resultRanges;
+  }
+
+  private boolean hasOverlappingAttendees(Event event, MeetingRequest request, Map<Event, Boolean> cache) {
+    Boolean isOverlapping = cache.get(event);
+    if (isOverlapping == null) {
+      isOverlapping = hasOverlappingAttendees(event, request);
+      cache.put(event, isOverlapping);
+    }
+    return isOverlapping;
```

Comment:
I think this would be more clear if you did:
```
if (cache.containsKey(event)) {
  return cache.get(event);
}
isOverlapping = hasOverlappingAttendees(event, request);
cache.put(event, isOverlapping);
```